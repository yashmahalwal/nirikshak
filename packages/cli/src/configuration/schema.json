{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "$id": "nirikshak/configuration",
    "title": "Nirikshak Configuration",
    "description": "Schema for nirikshak configuration",
    "type": "object",
    "properties": {
        "app": {
            "description": "Path to the API server",
            "type": "string",
            "minLength": 1
        },
        "dir": {
            "description": "Directory to be used for keeping tests",
            "type": "string",
            "minLength": 1
        },
        "resources": {
            "description": "Array of resources to be tested",
            "type": "array",
            "minItems": 1,
            "uniqueItems": true,
            "items": {
                "description": "Identifier for the resource, optionally mapped with directory",
                "type": ["string", "object"],
                "minLength": 1,
                "properties": {
                    "name": {
                        "description": "Name of the resource",
                        "type": "string",
                        "minLength": 1
                    },
                    "dir": {
                        "description": "Directory to be used for this resource (inside nikikshak directory)",
                        "type": "string",
                        "minLength": 1
                    }
                },
                "required": ["name", "dir"]
            }
        },
        "jestArgs": {
            "description": "Arguments to be passed to jest at runtime",
            "type": "array",
            "items": {
                "type": "string"
            }
        }
    },
    "required": ["dir", "resources", "app"]
}
