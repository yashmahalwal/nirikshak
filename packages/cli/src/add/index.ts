import { CliArgs } from "../utils/types";
import addToConfig from "./addToConfig";
import addToJest from "./addToJest";
import addDirectory from "./addDirectory";
import signale from "signale";
import { checkExistanceValidity } from "./checkExistanceValidity";

interface AddArgs extends CliArgs {
    name: string;
    dir?: string;
}

async function add({
    name,
    dir,
    config,
    configuration,
}: AddArgs): Promise<void> {
    // The directory entry for the resource
    const directory = dir ?? name;
    signale.info(`Adding resource ${name}`);
    // Check for any conflicts
    await checkExistanceValidity(name, directory, configuration);
    // Adding the resource directory and files for it
    await addDirectory(directory, configuration.dir, name, configuration.app);
    // Adding the entry to nirkshak configuration
    await addToConfig(name, config, dir);
    // Adding the entry to jest configuration
    await addToJest(name, configuration.dir, dir);
    signale.success("Done");
}

export const command = "add <name> [dir]";
export const describe = "Add a resource to the project";
export const handler = add;
