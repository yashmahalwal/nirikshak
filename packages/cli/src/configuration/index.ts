import path from "path";
import fs from "fs-extra";
import _ from "lodash";
import schema from "./schema.json";
import Ajv from "ajv";
import { Configuration } from "../utils/types";
import { getResourceName } from "../utils/resourceData";
import { getResourceDirectory } from "../utils/getResourceDir";

export const validateConfig = (data: any): void => {
    const ajv = new Ajv();
    // Validate with the stored JSON schema
    if (!ajv.validate(schema, data))
        throw `Error parsing configuration: ${ajv.errorsText()}`;

    const resources = (data as Configuration).resources.map((v) =>
        getResourceName(v)
    );
    // Check for resource uniqueness
    if (_.uniq(resources).length !== resources.length)
        throw `Duplicate resource names in configuration`;

    const directories = (data as Configuration).resources.map((v) =>
        path.resolve(getResourceDirectory(v, (data as Configuration).dir))
    );
    // Check for directory uniqueness
    if (_.uniq(directories).length !== resources.length)
        throw `Duplicate resource directories in configuration`;
};

export function parseConfig(
    configFile: string
): { configuration: Configuration } | Error {
    try {
        const data = fs.readJSONSync(configFile);
        validateConfig(data);
        return { configuration: data };
    } catch (e) {
        // Yargs requires an error object to be returned
        return new Error(e);
    }
}
