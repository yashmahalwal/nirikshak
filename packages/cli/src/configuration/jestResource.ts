import { CliArgs } from "../utils/additionalTypes";
import { JestConfig } from ".";

export function makeResourceJestConfiguration(
    resource: CliArgs["configuration"]["resources"][0]
): JestConfig {
    return {
        displayName: typeof resource == "string" ? resource : resource.name,
    };
}
