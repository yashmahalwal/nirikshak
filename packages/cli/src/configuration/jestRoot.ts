import { CliArgs } from "../utils/additionalTypes";
import { JestConfig } from ".";
import { PathJoiner } from "../utils/pathJoiner";
export function makeRootJestConfiguration(
  configuration: CliArgs["configuration"]
): JestConfig {
  const joiner = new PathJoiner();
  return {
    displayName: "all",
    projects: configuration.resources.map((resource) =>
      joiner.join(
        "<rootDir>",
        typeof resource == "string" ? resource : resource.dir
      )
    ),
  };
}
