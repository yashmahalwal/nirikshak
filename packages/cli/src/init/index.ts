import { Configuration } from "../utils/types";
import { ensureDirectories } from "./ensureDirectories";
import placeJestConfig from "./jest";
import signale from "signale";

async function init({
    configuration,
}: {
    configuration: Configuration;
}): Promise<void> {
    signale.info("Initialising nirikshak");
    // Prepare the folder structure needed
    await ensureDirectories(configuration);
    // Prepare jest files
    await placeJestConfig(configuration);
    // Adding entries for the resources in .nirikshak
    signale.success("Done");
}

export const command = "init";
export const describe = "Initialise nirikshak in the project";
export const handler = init;
