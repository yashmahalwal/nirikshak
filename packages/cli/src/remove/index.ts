import { CliArgs } from "../utils/types";
import removeDirectory from "./removeDirectory";
import removeFromConfig from "./removeFromConfig";
import removeFromJest from "./removeFromJest";
import signale from "signale";
import { getResourceName } from "../utils/resourceData";

interface RemoveArgs extends CliArgs {
    name: string;
}

async function remove({
    name,
    config,
    configuration,
}: RemoveArgs): Promise<void> {
    signale.info(`Removing resource ${name}`);
    // Check if the resource exists in the configuration
    if (!configuration.resources.some((e) => getResourceName(e) === name))
        throw new Error(`Cannot remove non existing resource: ${name}`);

    // Resource existance guaranteed
    // Remove the directory for it
    await removeDirectory(configuration, name);
    // Remove from nirikshak config
    await removeFromConfig(name, config);
    // Remove from jest config
    await removeFromJest(name);
    signale.success(`Done`);
}

export const command = "remove <name>";
export const describe = "Remove a resource from the project";
export const handler = remove;
