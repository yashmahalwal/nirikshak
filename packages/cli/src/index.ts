#!/usr/bin/env node
/* istanbul ignore file */
import figlet from "./utils/figletWrapper";
import chalk from "chalk";
import yargs from "yargs";
import * as Init from "./init";
import { parseConfig } from "./configuration";
import signale from "signale";
import * as Remove from "./remove";
import * as Add from "./add";
import * as Run from "./run";
/* Main file to route the commands to appropriate functions */

(async function (): Promise<void> {
    console.log(
        chalk.whiteBright(
            await figlet("Nirikshak", {
                horizontalLayout: "full",
            })
        )
    );
    console.log(chalk.underline.bold("Test your REST APIs with ease"));
    console.log("\n");

    yargs
        .demandCommand(1)
        // Init command
        .command(Init)
        // Add command
        .command(Add)
        // Remove command
        .command(Remove)
        // Run command
        .command(Run)
        // Update app command
        .usage("Usage: $0 <command> [options]")
        // Option: config
        .config("config", parseConfig)
        // Configuration file options
        .alias("config", "c")
        .default("config", "nirikshak.json")
        .describe("config", "path to the configuration file")
        // Examples
        .example(
            "$0 init",
            "Initialises the project using configuration in nirikshak.json"
        )
        .example(
            "$0 init -c config.json",
            "Initialises the project using configuration in config.json"
        )
        .example(
            "$0 add student",
            "Adds files for a resource named student in student directory."
        )
        .example(
            "$0 add student myDir",
            "Adds files for a resource named student in myDir directory."
        )
        .example("$0 remove student", "Remove the resource named student.")
        // Alias help
        .help()
        .alias("help", "h")
        // Alias version
        .version()
        .alias("version", "v")
        // Error handling
        .fail((message, error) => {
            if (message && message !== error?.message) {
                yargs.showHelp();
                signale.fatal(message);
            } else signale.fatal(error);
            process.exit();
        })
        .recommendCommands()
        // Start
        .parse();
})();
