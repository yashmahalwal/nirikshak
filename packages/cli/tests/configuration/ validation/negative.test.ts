import process from "process";
import { validateConfig } from "../../../src/configuration";
const invalidConfigs = [
    // No app
    {},
    // No app
    { dir: "testDirectory" },
    // No app
    { resources: ["student"] },
    // No app
    { dir: 12, resources: ["name"] },
    // No app
    { dir: "testDir", resources: 12312 },
    // No app
    { dir: "testDir", resources: [] },
    // No app
    { dir: "testDir", resources: [""] },
    // No app
    { dir: "testDir", resources: [{ name: "" }] },
    // No app
    { dir: "testDir", resources: [{ name: "name" }] },
    // No app
    { dir: "testDir", resources: [{ name: "name", dir: "" }] },
    // No app
    { dir: "testDir", resources: [{ name: "name", dir: "dir" }, ""] },
    // No app
    { dir: "testDir", resources: ["", { name: "" }] },
    // No app
    { dir: "testDir", resources: ["", { name: "name" }] },
    // No app
    { dir: "testDir", resources: ["", { name: "name", dir: "" }] },
    // No dir
    { app: "app.ts" },
    // No resource
    { app: "app.ts", dir: "testDirectory" },
    // No dir
    { app: "app.ts", resources: ["student"] },
    // Invalid dir
    { app: "app.ts", dir: 12, resources: ["name"] },
    // Invalid resources
    { app: "app.ts", dir: "testDir", resources: 12312 },
    // Invalid resource length
    { app: "app.ts", dir: "testDir", resources: [] },
    // Invalid resource entry
    { app: "app.ts", dir: "testDir", resources: [""] },
    // Invalid resource entry name
    { app: "app.ts", dir: "testDir", resources: [{ name: "" }] },
    // Missing dir from resource entry
    { app: "app.ts", dir: "testDir", resources: [{ name: "name" }] },
    // Invalid dir in resource entry
    { app: "app.ts", dir: "testDir", resources: [{ name: "name", dir: "" }] },
    // Invalid resource entry
    {
        app: "app.ts",
        dir: "testDir",
        resources: [{ name: "name", dir: "dir" }, ""],
    },
    // Invalid resource entry
    { app: "app.ts", dir: "testDir", resources: ["", { name: "" }] },
    // Invalid resource entry
    { app: "app.ts", dir: "testDir", resources: ["", { name: "name" }] },
    // Invalid resource entry
    {
        app: "app.ts",
        dir: "testDir",
        resources: ["", { name: "name", dir: "" }],
    },
    // Duplicate resource entry
    {
        app: "app.ts",
        dir: "testDir",
        resources: ["name", { name: "name", dir: "files" }],
    },
    // Duplicate resource directory
    {
        app: "app.ts",
        dir: "testDir",
        resources: ["name", { name: "namaste", dir: "name" }],
    },
    // Invalid jest argument
    {
        app: "app.ts",
        dir: "testDir",
        resources: ["name"],
        jestArgs: [1, 2, 3, 4, false],
    },
    // Invalid jestArgs type
    {
        app: "app.ts",
        dir: "testDir",
        resources: ["name"],
        jestArgs: "shingles",
    },
];

describe("Validate invalid config", () => {
    beforeAll(() => process.chdir(__dirname));
    test.each(invalidConfigs)("Validate invalid config %#", (invalid) => {
        expect.assertions(1);
        try {
            validateConfig(invalid);
        } catch (e) {
            expect(e).toMatchSnapshot();
        }
    });
});
