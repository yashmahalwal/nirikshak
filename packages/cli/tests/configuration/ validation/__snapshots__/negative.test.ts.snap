// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate invalid config Validate invalid config 0 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 1 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 2 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 3 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 4 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 5 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 6 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 7 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 8 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 9 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 10 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 11 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 12 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 13 1`] = `"Error parsing configuration: data should have required property 'app'"`;

exports[`Validate invalid config Validate invalid config 14 1`] = `"Error parsing configuration: data should have required property 'dir'"`;

exports[`Validate invalid config Validate invalid config 15 1`] = `"Error parsing configuration: data should have required property 'resources'"`;

exports[`Validate invalid config Validate invalid config 16 1`] = `"Error parsing configuration: data should have required property 'dir'"`;

exports[`Validate invalid config Validate invalid config 17 1`] = `"Error parsing configuration: data.dir should be string"`;

exports[`Validate invalid config Validate invalid config 18 1`] = `"Error parsing configuration: data.resources should be array"`;

exports[`Validate invalid config Validate invalid config 19 1`] = `"Error parsing configuration: data.resources should NOT have fewer than 1 items"`;

exports[`Validate invalid config Validate invalid config 20 1`] = `"Error parsing configuration: data.resources[0] should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 21 1`] = `"Error parsing configuration: data.resources[0].name should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 22 1`] = `"Error parsing configuration: data.resources[0] should have required property 'dir'"`;

exports[`Validate invalid config Validate invalid config 23 1`] = `"Error parsing configuration: data.resources[0].dir should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 24 1`] = `"Error parsing configuration: data.resources[1] should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 25 1`] = `"Error parsing configuration: data.resources[0] should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 26 1`] = `"Error parsing configuration: data.resources[0] should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 27 1`] = `"Error parsing configuration: data.resources[0] should NOT be shorter than 1 characters"`;

exports[`Validate invalid config Validate invalid config 28 1`] = `"Duplicate resource names in configuration"`;

exports[`Validate invalid config Validate invalid config 29 1`] = `"Duplicate resource directories in configuration"`;

exports[`Validate invalid config Validate invalid config 30 1`] = `"Error parsing configuration: data.jestArgs[0] should be string"`;

exports[`Validate invalid config Validate invalid config 31 1`] = `"Error parsing configuration: data.jestArgs should be array"`;
